components/brave_ads/browser/ads_p3a.cc

```diff
@@ -0,0 +1,147 @@
+}
+
+void UpdateMetricCount(PrefService* local_state,
+                       const std::string& pref_key,
+                       const std::string& histogram_name) {
+  DCHECK(local_state != nullptr);
+
+  DictionaryPrefUpdate update(local_state, kNewTabPageEventCountDictPref);
+  int count = 0;
+
+  base::Value* curr_value = update->FindPath({histogram_name, pref_key});
+  if (curr_value != nullptr && curr_value->is_int()) {
+    count = curr_value->GetInt();
+  }
+
+  count++;
+  update->SetPath({histogram_name, pref_key}, base::Value(count));
+  base::Time new_expiry_time = base::Time::Now() + kCountExpiryTime;
+  update->SetPath({histogram_name, kExpireTimeKey},
+                  base::Value(new_expiry_time.ToDoubleT()));
+  p3a_utils::RecordToHistogramBucket(histogram_name.c_str(), kCountBuckets,
+                                     count);
+}
+
+void DeleteOldCounts(PrefService* local_state) {
+  DCHECK(local_state != nullptr);
+
+  const base::Value::Dict& dict =
+      local_state->GetDictionary(kNewTabPageEventCountDictPref)->GetDict();
+
+  std::vector<std::string> delete_list;
+
+  for (auto it : dict) {
+    absl::optional<double> expiry_time =
+        it.second.FindDoubleKey(kExpireTimeKey);
+    if (base::Time::Now() >= base::Time::FromDoubleT(expiry_time.value_or(0))) {
+      delete_list.push_back(it.first);
+    }
+  }
+
+  if (!delete_list.empty()) {
+    DictionaryPrefUpdate update(local_state, kNewTabPageEventCountDictPref);
+    for (const auto& histogram_name : delete_list) {
+      update->RemovePath(histogram_name);
+    }
+  }
+}
+
+}  // namespace
+
+void RegisterLocalStatePrefs(PrefRegistrySimple* registry) {
+  registry->RegisterDictionaryPref(kNewTabPageEventCountDictPref);
+}
+
+void RecordNewTabPageEvent(PrefService* local_state,
+                           brave::BraveP3AService* p3a_service,
+                           const std::string& creative_instance_id,
+                           const ads::mojom::NewTabPageAdEventType event_type) {
+  DCHECK(local_state != nullptr);
+
+  DeleteOldCounts(local_state);
+
+  base::Time histogram_expiry_time = base::Time::Now() + kHistogramExpiryTime;
+
+  switch (event_type) {
+    case ads::mojom::NewTabPageAdEventType::kServed:
+    case ads::mojom::NewTabPageAdEventType::kViewed: {
+      p3a_service->RegisterMetric(
+          GetHistogramName(creative_instance_id,
+                           ads::mojom::NewTabPageAdEventType::kViewed),
+          brave::MetricLogType::kExpress, histogram_expiry_time);
+      p3a_service->RegisterMetric(
+          GetHistogramName(creative_instance_id,
+                           ads::mojom::NewTabPageAdEventType::kClicked),
+          brave::MetricLogType::kExpress, histogram_expiry_time);
+
+      UpdateMetricCount(local_state, kViewCountKey,
+                        GetHistogramName(creative_instance_id, event_type));
+      VLOG(1) << "AdsP3A: recorded viewed event for creative: "
+              << creative_instance_id;
```

Comment:
We should remove most logging calls before checking in. See https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md#logging